Objects/typeobject.c

```diff
@@ -4759,20 +4965,59 @@ _PyType_Lookup(PyTypeObject *type, PyObject *name)
     unsigned int h = MCACHE_HASH_METHOD(type, name);
     struct type_cache *cache = get_type_cache();
     struct type_cache_entry *entry = &cache->hashtable[h];
+#ifdef Py_GIL_DISABLED
+    // synchronize-with other writing threads by doing an acquire load on the sequence
+    while (1) {
+        int sequence = _Py_atomic_load_int_acquire(&entry->sequence);
+        if (!TYPE_CACHE_IS_UPDATING(sequence)) {
+            if (_Py_atomic_load_uint32_relaxed(&entry->version) == type->tp_version_tag &&
+                _Py_atomic_load_ptr_relaxed(&entry->name) == name) {
+                assert(_PyType_HasFeature(type, Py_TPFLAGS_VALID_VERSION_TAG));
+                OBJECT_STAT_INC_COND(type_cache_hits, !is_dunder_name(name));
+                OBJECT_STAT_INC_COND(type_cache_dunder_hits, is_dunder_name(name));
+                PyObject *value = _Py_atomic_load_ptr_relaxed(&entry->value);
+
+                // Synchronize again and validate that the entry hasn't been updated
+                // while we were readying the values.
+                if (_Py_atomic_load_int_acquire(&entry->sequence) == sequence) {
+                    return value;
+                }
+             } else {
+                // Cache miss
+                break;
+             }
+        }
+        // We are in progress of updating the cache, retry
+        _Py_yield();
+    }
+#else
     if (entry->version == type->tp_version_tag &&
         entry->name == name) {
         assert(_PyType_HasFeature(type, Py_TPFLAGS_VALID_VERSION_TAG));
         OBJECT_STAT_INC_COND(type_cache_hits, !is_dunder_name(name));
         OBJECT_STAT_INC_COND(type_cache_dunder_hits, is_dunder_name(name));
         return entry->value;
     }
+#endif
     OBJECT_STAT_INC_COND(type_cache_misses, !is_dunder_name(name));
     OBJECT_STAT_INC_COND(type_cache_dunder_misses, is_dunder_name(name));
 
     /* We may end up clearing live exceptions below, so make sure it's ours. */
     assert(!PyErr_Occurred());
 
+    // We need to atomically do the lookup and capture the version before
+    // anyone else can modify our mro or mutate the type.
+
+    int has_version = 0;
+    int version = 0;
+    BEGIN_TYPE_LOCK()
     res = find_name_in_mro(type, name, &error);
+    if (MCACHE_CACHEABLE_NAME(name)) {
+        has_version = assign_version_tag(interp, type);
+        version = type->tp_version_tag;
+    }
+    END_TYPE_LOCK()
```

Comment:
Should we be updating the cache below while still holding type lock?

Reply:
It's fine that we're not doing that because we're capturing the version here with the lock held.  If something mutates the lock after this then we'll just stuff an invalid version entry into the cache versus having the cache entry get invalidated right after we release the lock.